nixos/privoxy: add https and settings options

rnhmjoj opened the pull request:
###### Motivation for this change

This is a major rewrite of the Privoxy module:

- As per RFC0042, remove privoxy.extraConfig and replace it
  with a privoxy.settings option, which maps a NixOS freeform
  submodule to the Privoxy configuration format.

- Move all top-level options that mirrored a setting to
  the real ones in privoxy.settings. This still keeps the
  type-checking, default values and examples in places.

- Add two convenience options: userActions and userFilters, which
  simplify the operation of creating a file with pkgs.writeText,
  converting it to a string and adding it to the actionsfile/
  filterfile list.

- Add a privoxy.inspectHttps option to automagically setup TLS
  decryption support. I don't know how long have been waiting
  for this feature: can't believe it has just happened.

- Also add a privoxy.certsLifetime to control the periodical
  cleanup of the temporary certificates generate by Privoxy.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via new `nixosTests.privoxy`
- [x] Tested compilation of all pkgs that depend on this change
- [x] Tested execution of all binary files
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
privoxy: enable more useful features

This cause ~6MB of closure size increase.

rnhmjoj committed:
nixos/tests/privoxy: add test

rnhmjoj committed:
nixos/privoxy: add https and settings options

This is a major rewrite of the Privoxy module:

- As per RFC0042, remove privoxy.extraConfig and replace it
  with a privoxy.settings option, which maps a NixOS freeform
  submodule to the Privoxy configuration format.

- Move all top-level options that mirrored a setting to
  the real ones in privoxy.settings. This still keeps the
  type-checking, default values and examples in places.

- Add two convenience options: userActions and userFilters, which
  simplify the operation of creating a file with pkgs.writeText,
  converting it to a string and adding it to the actionsfile/
  filterfile list.

- Add a privoxy.inspectHttps option to automagically setup TLS
  decryption support. I don't know how long have been waiting
  for this feature: can't believe it has just happened.

- Also add a privoxy.certsLifetime to control the periodical
  cleanup of the temporary certificates generate by Privoxy.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 115005` at 6b6a20d7 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/7484d1ffaa73a9366af32aab34c96868">privoxy</a></li></li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/tools/networking/privoxy/default.nix:
```diff
@@ -13,14 +18,17 @@ stdenv.mkDerivation rec {
+  configureFlags = [ "--with-mbedtls" "--with-brotli" ];
 
   postInstall = ''
     rm -rf $out/var
```
```suggestion
    rm -r $out/var
```

rnhmjoj committed:
nixos/privoxy: set temporary directory

This is needed for working external filters, otherwise privoxy will fail
without a clear error message.

rnhmjoj committed:
nixos/release-notes: highlight privoxy updates

emmanuelrosa commented:
How can I define multiple `forward` definitions? I've tried to migrate my config as follows:

```
    settings = {
      toggle = true;
      forward = "/  .";
      forward-socks5t = ".onion/ 127.0.0.1:9050 .";
      forward = "192.168.*.*/  .";
      forward = "10.*.*.*/  .";
      forward = "127.*.*.*/  .";
      forward = "localhost/  .";
    };
```

The problem is that it doesn't support more than one `forward` definition.

emmanuelrosa commented:
OK, so that would translate into this:

```
    settings = {
      toggle = true;
      forward = [ "/  ."
                ];
      forward-socks5t = ".onion/ 127.0.0.1:9050 .";
      forward = [ "192.168.*.*/  ."
                  "10.*.*.*/  ."
                  "127.*.*.*/  ."
                  "localhost/  ."
                ];
    };
```

So there would still be two `forward` definitions, right?

rnhmjoj commented:
Sorry but I don't understand what you're trying to achieve: is the order important?
Can't you just join the two `forward` definitions?

emmanuelrosa commented:
Upon closer inspection, while the order does matter, in my case I should be able to put the `forward-socks5t` rule at the end, which leads to this configuration:

```

  services.privoxy = {
    enable = true;
    settings = {
      toggle = true;
      forward = [ " /                    ."
                  "192.168.*.*/          ."
                  "10.*.*.*/             ."
                  "127.*.*.*/            ."
                  "localhost/            ."
                ];
      forward-socks5t = ".onion/              127.0.0.1:9050 .";
    };   
  };
```

However, it fails to build with the error:

```
error: undefined variable 'certificate-directory' at /home/emmanuel/projects/nixpkgs/nixos/modules/services/networking/privoxy.nix:195:12
```

It's unclear what `settings.certificate-directory` should be set to, given I'm not using that feature.

rnhmjoj commented:
This is an actual mistake: I fixed it in 7962df46feb. I tested your configuration in VM and now it's working.
Thank you for reporting it.

emmanuelrosa commented:
Thank you, that worked.
